hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -363,8 +376,45 @@ class ServiceBackend() extends Backend {
   }
 
   def lowerDistributedSort(ctx: ExecuteContext, stage: TableStage, sortFields: IndexedSeq[SortField], relationalLetsAbove: Map[String, IR]): TableStage = {
-    // Use a local sort for the moment to enable larger pipelines to run
-    LowerDistributedSort.localSort(ctx, stage, sortFields, relationalLetsAbove)
+    val region = ctx.r
+    val rowType = stage.rowType
+    val keyType = rowType.typeAfterSelectNames(sortFields.map(_.field))
+    val rowEType = EType.fromTypeAndAnalysis(rowType, TypeWithRequiredness(rowType)).asInstanceOf[EBaseStruct] // FIXME(danking): Probably not kosher
+    val keyEType = EType.fromTypeAndAnalysis(keyType, TypeWithRequiredness(keyType)).asInstanceOf[EBaseStruct] // FIXME(danking): Probably not kosher
+    val shuffleType = TShuffle(sortFields, rowType, rowEType, keyEType)
+    val shuffleClient = new ShuffleClient(shuffleType, ctx)
+    assert(keyType == shuffleClient.codecs.keyType)
+    val keyDecodedPType = shuffleClient.codecs.keyDecodedPType
+    shuffleClient.start()
+    try {
+      val uuid = shuffleClient.uuid
+
+      val successfulPartitionIds = execute(
+        ctx,
+        stage.mapCollect(relationalLetsAbove)(
+          ShuffleWrite(Literal(shuffleType, uuid), _)))
+
+      val partitionBoundsPointers = shuffleClient.partitionBounds(region, stage.numPartitions) // FIXME(danking): number of partitions should be configurable?
+      val partitionIntervals = partitionBoundsPointers.zip(partitionBoundsPointers.drop(1)).map { case (l, r) =>
+        Interval(SafeRow(keyDecodedPType, l), SafeRow(keyDecodedPType, r), includesStart = true, includesEnd = false)
+      }
+      val last = partitionIntervals.last
+      partitionIntervals(partitionIntervals.length - 1) = Interval(
+        last.left.point, last.right.point, includesStart = true, includesEnd = true)
+
+      val partitioner = new RVDPartitioner(keyType, partitionIntervals.toFastIndexedSeq)
+
+      TableStage(
+        globals = Literal(TStruct(), Row()),
+        partitioner = partitioner,
+        TableStageDependency.none,
+        contexts = ToStream(MakeArray(
+          partitionIntervals.map(interval => Literal(TInterval(keyType), interval)),
+          TArray(TInterval(keyType)))),
+        interval => ShuffleRead(Literal(shuffleType, uuid), interval))
+    } finally {
```

Comment:
So the `finally` block closes the shuffleClient, but we'll still be able to read the result of this shuffle by creating a new shuffle client with the correct uuid, right?

Do we ever clean up the results of a shuffle once it's passed out of scope? I think it's pretty safe (for now) to delete a shuffle once we've exited the nested `execute` call on the IR that created it, since we can't reference it otherwise.

Reply:
Right, this closes the socket. ShuffleClient.stop frees the shuffle.

This is another question for @tpoterba: when is the `ShuffleRead` on line 414 no longer usable? That ShuffleRead is contained inside a `TableStage`. The `TableStage` is returned by lowerDistributedSort. Is there any lifetime management I can piggyback on?

Reply:
You should call `shuffleClient.close()` when the ExecuteContext is closed, I think.